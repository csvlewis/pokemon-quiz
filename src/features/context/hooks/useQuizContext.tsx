import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
