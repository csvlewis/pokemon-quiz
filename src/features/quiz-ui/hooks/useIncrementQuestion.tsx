import { useQuizContext } from "@/features/context/components/QuizContextProvider";

export function useIncrementQuestion(): () => void {
  const {
    currentQuestion,
    questions,
    setCurrentQuestion,
    setDisplayAnswer,
    setSelectedAnswer,
    setShowResults,
    setQuizInProgress,
  } = useQuizContext();

  return () => {
    if (currentQuestion >= questions.length) {
      setQuizInProgress(false);
      setShowResults(true);
      return;
    }
    setCurrentQuestion(currentQuestion + 1);
    setDisplayAnswer(false);
    setSelectedAnswer({});
  };
}
