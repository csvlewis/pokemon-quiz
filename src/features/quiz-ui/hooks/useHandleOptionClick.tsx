import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
