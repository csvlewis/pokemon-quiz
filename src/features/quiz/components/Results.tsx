import { Button, Paper, Stack, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function Results(): React.ReactNode {
  const { correctAnswers, questions, quizInProgress, resetQuiz, showResults } =
    useQuizContext();

  if (!quizInProgress && showResults)
    return (
      <Paper>
        <Stack alignItems="center" spacing={2} sx={{ margin: 2 }}>
          <Typography>
            You got {correctAnswers} out of {questions.length} answers correct
          </Typography>
          <Typography>Would you like to play again?</Typography>
          <Button variant="contained" onClick={() => resetQuiz()}>
            <Typography>Create New Quiz</Typography>
          </Button>
        </Stack>
      </Paper>
    );
}
