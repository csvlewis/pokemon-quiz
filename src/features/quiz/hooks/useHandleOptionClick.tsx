import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
