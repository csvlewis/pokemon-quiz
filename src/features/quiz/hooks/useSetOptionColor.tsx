import { OverridableStringUnion } from "@mui/types";

import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
