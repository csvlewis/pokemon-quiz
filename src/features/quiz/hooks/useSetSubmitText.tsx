import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetSubmitText(): () => string {
  const { currentQuestion, displayAnswer, questions } = useQuizContext();

  return () => {
    let text: string = "Submit";

    if (displayAnswer) {
      if (currentQuestion >= questions.length) {
        text = "See Results";
      } else text = "Next Question";
    }

    return text;
  };
}
