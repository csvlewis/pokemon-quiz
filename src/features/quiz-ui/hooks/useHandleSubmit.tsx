import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
