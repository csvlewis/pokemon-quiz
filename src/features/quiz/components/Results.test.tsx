import React from "react";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Results } from "./Results";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("Results", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      correctAnswers: 0,
      questions: [{}, {}],
      quizInProgress: false,
      resetQuiz: jest.fn(),
      showResults: true,
    });

    const { container } = render(<Results />);
    expect(container).toMatchSnapshot();
  });
});
