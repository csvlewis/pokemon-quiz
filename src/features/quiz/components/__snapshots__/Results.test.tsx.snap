// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results should render correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiStack-root css-d32kqm-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        You got 
        0
         out of 
        2
         answers correct
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Would you like to play again?
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          Create New Quiz
        </p>
      </button>
    </div>
  </div>
</div>
`;
